ext/msgpack/unpacker.c

```diff
@@ -160,6 +179,16 @@ static inline int object_complete_binary(msgpack_unpacker_t* uk, VALUE str)
 #ifdef COMPAT_HAVE_ENCODING
     ENCODING_SET(str, msgpack_rb_encindex_ascii8bit);
 #endif
+
+#if STR_UMINUS_DEDUPE
+    if(uk->freeze) {
+# if STR_UMINUS_DEDUPE_FROZEN
+        rb_str_freeze(str);
```

Comment:
Please don't merge just yet, this part is not quite correct. I'll push a fix in a couple minutes.

Reply:
It's fixed now.

Reply:
I can't see the commit to fix this (but the whole commit forcibly pushed).
Could you explain what was wrong and fixed? I want to understand it.

Reply:
Of course. So the `String#-@` behavior changed a lot on every single version.

Up to 2.5, it had a simple guard statement `return self if frozen?`. Note that frozen doesn't mean "interned". So if you want to deduplicate a string in 2.5 or older, you have to supply a mutable string.

2.6 and 2.7 are a bit smarter, and will deduplicate a supplied string even if it is frozen, however the code is a bit like: `str = str.dup if str.frozen?`, meaning that calling `-@` on a frozen string doesn't give you anything compared to a mutable one.

Then starting in 2.8 (named 3.0 now), if the string is frozen it can directly be put into the interned string table, so freezing the string before calling `-@` will save you an allocation per unique string.

Now my feature checking condition I pushed was a bit broken. The intent was to test wether frozen string could be directly promoted to interned string.

Here's the first version of my test:

```ruby
  a = -(%w(t e s t).join.freeze)
  b = -(%w(t e s t).join.freeze)
  if a.equal?(b)
```

This check was buggy because on 2.6 and 2.7, both expression would return a third instance that was previously created an interned. No huge deal, but it mean we'd do one useless `rb_str_freeze` call for each string.

So I rewrote it as:

```ruby
  s = rand.to_s.freeze
  if (-s).equal?(s)
```

Which did properly work on 2.6+, however it was broken on 2.5- because while the returned string was the same instance, it wasn't interned at all (because `return self if frozen?`).

So I finally pushed:

```ruby
  s = rand.to_s.freeze
  if (-s).equal?(s) && (-s.dup).equal?(s)
```

Which properly check that frozen strings can be promoted to interned strings.

I hope I explained this properly.

Reply:
That makes sense, perfectly. Thanks!